rfc/0009-audit-checks.md

```diff
@@ -0,0 +1,171 @@
+As a result of that, a Kubernetes resource that was marked as valid even a couple of hours ago,
+could be instead rejected later on because of a policy change.
+
+We want to implement a audit check that inspects all the resources that are already
+present inside of the cluster and flags all the ones that are violating at least one of
+the currently enforced policies.
+
+## Examples / User Stories
+
+As a operator, I want to know if some resource in my cluster violates the latest policies.
+
+As a user, I want to know if some of resource that I own violates the cluster compliance policies.
+
+As a Kubernetes developer, I want access to the audit checks results to allow me pro-grammatically act upon it.
+
+# Detailed design
+
+The audit checks will require changes in `PolicyServer`, policies and the creation of a new component `Background Audit Scanner`.
+
+## Policy server
+
+`PolicyServer` needs to store successful evaluation from all policies, and also failed evaluations for 
+policies in monitor mode. Failed evaluations for policies in enforce mode will not be stored as resources won't be 
+created in the cluster. These evaluations will be stored in a `PolicyReport`. There is one `PolicyReport` per namespace and 
+one `ClusterPolicyReport` for cluster-wide resources, see this [RFC](https://github.com/kubewarden/rfc/pull/13/files) for more info.
+
+Many more calls will be made to `PolicyServer` for the background checks, therefore we need to implement a cache
+for policy evaluations. The result of each policy evaluation will be stored in the cache, it will contain the `resourceVersion`
+of the policy and the resource to see if any of them have changed. Before evaluating a new resource it will check in the 
+cache if evaluation was previously done. 
+
+## Background Audit Scanner
+
+By default, background audit scanning happens every hour, and it can be configured via the `background-scan` flag.
+It will scan all resources in the cluster for all policies.
+When a background audit has finished processing all evaluations it will replace all `PolicyReports` and the `ClusterPolicyReport`
+with the new content. This will remove old entries with resources that are deleted and make sure the reports are up to date. 
+
+Background Audit Scanner won't perform policy evaluation, it will make a request to `PolicyServer` for evaluating the resources
+instead.
+
+It should implement a caching mechanism to avoid overloading the Kubernetes api server. We can use client-go caching.
+
+A background audit checks consist of the following steps:
+1. Get information about which policies have audit checks enabled
+2. Get the resources configured in the policies with audit checks enabled
+3. Request the policy evaluation for all resources to the `PolicyServer`
+4. Aggregate all the policies validation results
+5. Publish a `PolicyReport` per namespace
+with the aggregated results and the `ClusterPolicyReport`.
```

Comment:
(As raised in [the other RFC](https://github.com/kubewarden/rfc/pull/13#discussion_r975215856))

- How atomically is a PolicyReport updated?
  - Does the PolicyReport contain all the evaluations for the *last run* of the audit of the namespace? (if results are present in the report, it is assumed that any result is the latest)
  - Or the *last run* of the resource? (if a result for resource is present in the report, it is assumed that the results are the latest for that specific resourse, but may not be for other resources)
- How frecuently are PolicyReportResults garbage-collected?
  - Do we only save the latest Result per (resource, policy) tuple?
  - Do we save older Results? How do we configure how many? on Kubewarden, or the PolicyReport itself?
    If we save older results, I fear it will not be efficient to try to find specific results for a resource.

Reply:
>How atomically is a PolicyReport updated? 
If both PolicyServer and Background Auditor Checks writes to the same PolicyReport we would need to implement some retry mechanism as it says [here](https://github.com/kubewarden/rfc/blob/audit-checks/rfc/0009-audit-checks.md?plain=1#L140-L142)

>How frecuently are PolicyReportResults garbage-collected? 

https://github.com/kubewarden/rfc/blob/audit-checks/rfc/0009-audit-checks.md?plain=1#L58-L59

Reply:
The approaches then contradict each other:
- 1. [Retry adding results to the PolicyReport](https://github.com/kubewarden/rfc/blob/audit-checks/rfc/0009-audit-checks.md?plain=1#L140-L142). So we are updating the PolicyReport all the time during the audit run.
- 2. [Wait until audit run is done to update the PolicyReport](https://github.com/kubewarden/rfc/blob/audit-checks/rfc/0009-audit-checks.md?plain=1#L58-L59).


> If both PolicyServer and Background Auditor Checks writes to the same PolicyReport we would need to implement some retry mechanism as it says [here](https://github.com/kubewarden/rfc/blob/audit-checks/rfc/0009-audit-checks.md?plain=1#L140-L142)

In my opinion we are wavering a bit the problem here. How do we implement this? naively retrying?  And what happens when there's a success in adding a new Result? The resource version changes, hence other retries may or may not be valid? Are we doing append-only then? If so, we should document that PolicyReport is append-only, so we can retry. And later on will be garbage-collected on number or entries or timestamp.

> https://github.com/kubewarden/rfc/blob/audit-checks/rfc/0009-audit-checks.md?plain=1#L58-L59

But, right now we are saying that PolicyServers create/update/delete things in the PolicyReports. How does one PolicyServer know that the other PolicyServers have finished evaluating all requests of the current audit run? I don't think this is easy. Hence https://github.com/kubewarden/rfc/blob/audit-checks/rfc/0009-audit-checks.md?plain=1#L58-L59 is not easily implemented.

Apart from the implementation, as a user and consumer of the PolicyReports,  knowing that the PolicyReport contains all the full evaluations for the *last audit run* is more valuable to me. Plus, as Kubewarden devs, I think we can promise this.

Reply:
PolicyServers just append new PolicyReports when evaluation happens (never in the background check). These evaluations will be removed when the background check happens, so after a background check we know that all resources have been fully scanned and the latest status of the cluster.

Background Auditor Check make all requests to PolicyServer, so it knows when all evaluation have finished and it then  updates the PolicyReport (it removes all previous records and updates with the latest cluster status).

I'm thinking now that it might be better to just do the background scan for now, and not store any evaluations in the PolicyServer. We would lose some of the real time info, but anyway the real time is not accurate because we are not tracking all events. I will add this, and the dry-run as other options to the RFC.